File: src/Test/RBTreeFFI.go
package Test_RBTreeFFI

type color bool

const (
	red   color = false
	black color = true
)

type tree struct {
	color color
	left  *tree
	value int
	right *tree
}

var pool []tree
var poolIdx int

func alloc(c color, l *tree, v int, r *tree) *tree {
	pool[poolIdx] = tree{color: c, left: l, value: v, right: r}
	p := &pool[poolIdx]
	poolIdx++
	return p
}

func balance(c color, a *tree, x int, b *tree) *tree {
	if c == black {
		if a != nil && a.color == red {
			if a.left != nil && a.left.color == red {
				return alloc(red, alloc(black, a.left.left, a.left.value, a.left.right), a.value, alloc(black, a.right, x, b))
			}
			if a.right != nil && a.right.color == red {
				return alloc(red, alloc(black, a.left, a.value, a.right.left), a.right.value, alloc(black, a.right.right, x, b))
			}
		}
		if b != nil && b.color == red {
			if b.left != nil && b.left.color == red {
				return alloc(red, alloc(black, a, x, b.left.left), b.left.value, alloc(black, b.left.right, b.value, b.right))
			}
			if b.right != nil && b.right.color == red {
				return alloc(red, alloc(black, a, x, b.left), b.value, alloc(black, b.right.left, b.right.value, b.right.right))
			}
		}
	}
	return alloc(c, a, x, b)
}

func ins(x int, t *tree) *tree {
	if t == nil {
		return alloc(red, nil, x, nil)
	}
	if x < t.value {
		return balance(t.color, ins(x, t.left), t.value, t.right)
	} else if x > t.value {
		return balance(t.color, t.left, t.value, ins(x, t.right))
	}
	return t
}

func insert(x int, t *tree) *tree {
	res := ins(x, t)
	return alloc(black, res.left, res.value, res.right)
}

func buildTree(n int, acc *tree) *tree {
	pool = make([]tree, 10000000) 
	poolIdx = 0

	for i := n; i > 0; i-- {
		acc = insert(i, acc)
	}
	return acc
}

func depth(t *tree) int {
	if t == nil {
		return 0
	}
	ld := depth(t.left)
	rd := depth(t.right)
	if ld > rd {
		return 1 + ld
	}
	return 1 + rd
}

func RunRBTreeFFI(limit float64) float64 {
	t := buildTree(int(limit), nil)
	return float64(depth(t))
}

File: src/Test/RBTreeFFI.rs
use std::rc::Rc;

#[derive(Clone, Copy, PartialEq)]
enum Color {
    Red,
    Black,
}

enum Tree {
    Leaf,
    Node(Color, Rc<Tree>, i64, Rc<Tree>),
}

fn alloc(c: Color, l: Rc<Tree>, v: i64, r: Rc<Tree>) -> Rc<Tree> {
    Rc::new(Tree::Node(c, l, v, r))
}

fn balance(c: Color, a: Rc<Tree>, x: i64, b: Rc<Tree>) -> Rc<Tree> {
    if c == Color::Black {
        if let Tree::Node(Color::Red, al, a_val, ar) = &*a {
            if let Tree::Node(Color::Red, all, all_val, alr) = &**al {
                return alloc(Color::Red, alloc(Color::Black, all.clone(), *all_val, alr.clone()), *a_val, alloc(Color::Black, ar.clone(), x, b.clone()));
            }
            if let Tree::Node(Color::Red, arl, arl_val, arr) = &**ar {
                return alloc(Color::Red, alloc(Color::Black, al.clone(), *a_val, arl.clone()), *arl_val, alloc(Color::Black, arr.clone(), x, b.clone()));
            }
        }
        if let Tree::Node(Color::Red, bl, b_val, br) = &*b {
            if let Tree::Node(Color::Red, bll, bll_val, blr) = &**bl {
                return alloc(Color::Red, alloc(Color::Black, a.clone(), x, bll.clone()), *bll_val, alloc(Color::Black, blr.clone(), *b_val, br.clone()));
            }
            if let Tree::Node(Color::Red, brl, brl_val, brr) = &**br {
                return alloc(Color::Red, alloc(Color::Black, a.clone(), x, bl.clone()), *b_val, alloc(Color::Black, brl.clone(), *brl_val, brr.clone()));
            }
        }
    }
    alloc(c, a, x, b)
}

fn ins(x: i64, t: &Rc<Tree>) -> Rc<Tree> {
    match &**t {
        Tree::Leaf => alloc(Color::Red, Rc::new(Tree::Leaf), x, Rc::new(Tree::Leaf)),
        Tree::Node(c, l, y, r) => {
            if x < *y {
                balance(*c, ins(x, l), *y, r.clone())
            } else if x > *y {
                balance(*c, l.clone(), *y, ins(x, r))
            } else {
                t.clone()
            }
        }
    }
}

fn insert(x: i64, t: &Rc<Tree>) -> Rc<Tree> {
    let res = ins(x, t);
    if let Tree::Node(_, l, y, r) = &*res {
        alloc(Color::Black, l.clone(), *y, r.clone())
    } else {
        Rc::new(Tree::Leaf)
    }
}

fn build_tree(n: i64, mut acc: Rc<Tree>) -> Rc<Tree> {
    let mut i = n;
    while i > 0 {
        acc = insert(i, &acc);
        i -= 1;
    }
    acc
}

fn depth(t: &Rc<Tree>) -> i64 {
    match &**t {
        Tree::Leaf => 0,
        Tree::Node(_, l, _, r) => {
            let ld = depth(l);
            let rd = depth(r);
            if ld > rd {
                1 + ld
            } else {
                1 + rd
            }
        }
    }
}

pub fn Test_RBTreeFFI_runRBTreeFFI(mut limit: i64) -> i64 {
    let t = build_tree(limit, Rc::new(Tree::Leaf));
    depth(&t)
}